### PR TITLE
[Statistics] Added sub-project filter to stats module

### PR DESCRIPTION
**Pull request by ZainVirani**
Behavioural/Reliability now have a sub-project filter

### CODE REVIEW

**Comment by ZainVirani** [comment]
[Redmine ticket #6822](https://redmine.cbrain.mcgill.ca/issues/6822#change-47726)

**Comment by sruthymathew123** [comment]
@ZainVirani @ridz1208 Tested on IBIS and working well.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 1]
```diff
@@ -545,19 +548,47 @@ class NDB_Form_Statistics extends NDB_Form
...
+        ) {
+            $currentSubProject = $_REQUEST['BehaviouralSubproject'];
+            $this->tpl_data['CurrentSubproject']
+                = array(
+                   'ID'   => $currentSubProject,
+                   'Name' => $subprojects[$currentSubProject],
+                  );
+            $suproject_query
+                = "AND (s.SubprojectID IS NULL".
+                " OR s.SubprojectID = ($currentSubProject))";
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: I didn't look over this code too closely, but at least this line is vulnerable to an SQL injection attack. `$currentSubProject` directly comes from the user, it should never be inserted into an SQL statement.

Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: @driusan `$currentSubProject` comes from `$_REQUEST['BehaviouralSubproject']` which comes from a drop down which only allows valid subprojects to be chosen
additionally i have
```
      if (isset($_REQUEST['BehaviouralSubproject'])
            && $_REQUEST['BehaviouralSubproject'] != ''
            && in_array($_REQUEST['BehaviouralSubproject'], array_keys($subprojects))
        ) {
```
which ensures the chosen option is valid
I basically follow the same method you wrote for choosing a valid Project
Unless I'm misunderstanding something, I don't see how the user would be able to set the value to something other than a valid subproject

File: modules/statistics/js/form_stats_behavioural.js [hunk 2]
```diff
@@ -1,10 +1,11 @@
 /*global document, $, window, scrollContent*/
 function updateBehaviouralTab() {
     var BehaviouralProject = document.getElementById("BehaviouralProject");
+    var BehaviouralSubproject = document.getElementById("BehaviouralSubproject");;
     var request            = $.ajax(
         {
-            url: loris.BaseURL + '/statistics/stats_behavioural/?dynamictabs=dynamictabs&BehaviouralProject=' + (BehaviouralProject==null ? "" : BehaviouralProject.value),
-            type: 'GET',
+            url: loris.BaseURL + '/statistics/stats_behavioural/?dynamictabs=dynamictabs&BehaviouralProject=' + (BehaviouralProject==null ? "" : BehaviouralProject.value) + '&BehaviouralSubproject=' + (BehaviouralSubproject==null ? "" : BehaviouralSubproject.value),
+            type: 'POST',
```
Comment (driusan) [inline modules/statistics/js/form_stats_behavioural.js]: Why is this changed to a POST? (Especially since the things that are added are GET parameters, not posted..)

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 1 above)
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: You're right, I missed the in_array, that does likely protect it from injection, but the fact that it's from a dropdown doesn't really matter. Hackers have access to curl.

Regardless of the in_array, it would still be far safer to use a prepared statement. I believe the project original was written before the prepared select wrappers existed in LORIS (and if it didn't, that should have been caught in review too). Any equality check in a `WHERE` clause should use a prepared parameters. The only exception is `IN` clauses, because you can't bind an array to a prepared statement variable (which is really annoying.)

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 3]
```diff
@@ -1064,17 +1138,22 @@ class NDB_Form_Statistics extends NDB_Form
...
+                $Param_Site
+            );
     }
     /**
      * GetReliabilityData function
      *
-     * @param string $Param_Project the value of Param_Project
-     * @param string $Param_Site    the value of Param_Site
+     * @param string $Param_Project   the value of Param_Project
+     * @param string $suproject_query the value of suproject_query
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: I think `suproject` is a typo, but as long as you're changing these can you also clean up the description comment for the other two? (Of course any variable has the value of the variable! But what's the meaning of the variable supposed to represent?)

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 4]
```diff
@@ -1113,15 +1193,17 @@ class NDB_Form_Statistics extends NDB_Form
     /**
      * GetSingleReliabilityData function
      *
-     * @param string $Param_Project the value of Param_Project
-     * @param string $Param_Site    the value of Param_Site
-     * @param string $name          the value of name
-     * @param string $inst_name     the value of inst_name
-     * @param string $cutoff        the value of cutoff
+     * @param string $Param_Project   the value of Param_Project
+     * @param string $suproject_query the value of suproject_query
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: same typo? (and comment about fixing the others as long as you're fixing this one.)

File: modules/statistics/js/form_stats_behavioural.js (hunk 2 above)
Comment (ZainVirani) [inline modules/statistics/js/form_stats_behavioural.js]: Oops, I remember messing with this but I forgot to change it back.

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 3 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Yeah I left `suproject` as it is just in case you did it for some specific reason. I can go through and make these changes. And I was just following your comment style for consistency ;) But yeah I'll run through all the relevant files and clean them up.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 5]
```diff
@@ -1029,6 +1060,49 @@ class NDB_Form_Statistics extends NDB_Form
...
+        } else {
+            $subproject_query ="";
+        }
+        if (!in_array(
+            $_REQUEST['BehaviouralSubproject'],
+            array_keys($subprojects)
+        )) {
+            $BS = 'BehaviouralSubproject';
+            $this->tpl_data['PreviousSubproject']
+                = Utility::getSubprojectsForProject()[$_REQUEST[$BS]];
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: maybe I'm missing something that defends against it here the same way I missed in_array above, but this looks like it's vulnerable to an XSS attack to me. The guard on this if statement is explicitly if it's _not_ in subprojects, so it doesn't protect malicious input from getting here.

(The value assigned to the template should pass through something that escapes it like htmlspecialchars to fix it.)

Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: If it's not in the array I do nothing with the data, I simply use the default drop down option to display all data for that project/all projects. But I see your point, I'll probably use htmlspecialchars.

**Comment by ZainVirani** [comment]
@driusan changes pushed; moving to Needs Testing

**Comment by jstirling91** [comment]
@ZainVirani assigning Dave to the pull request and putting it back to needs review

**Comment by driusan** [comment]
@johnsaigle can you review this from a security perspective? The $_REQUEST superglobal makes me a little suspicious.

**Comment by ridz1208** [comment]
@sruthymathew123 please re-test

**Comment by leothomas** [comment]
Filtering in reliability works correctly with IBIS data set. Are there any other modules, or corner cases to test? @ZainVirani @ridz1208

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 6]
```diff
@@ -545,19 +548,48 @@ class NDB_Form_Statistics extends NDB_Form
...
+                  );
+            $subproject_query     = "AND (s.SubprojectID IS NULL";
+            $subproject_query    .= " OR s.SubprojectID=:spid)";
+            $this->params['spid'] =$currentSubProject;
         } else {
-            $suproject_query ='';
+            $subproject_query ="";
+        }
+        if (!in_array(
+            $_REQUEST['BehaviouralSubproject'],
```
Comment (ridz1208) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: please add the "true" argument as a third parameter for the in_array function

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 7]
```diff
@@ -545,19 +548,48 @@ class NDB_Form_Statistics extends NDB_Form
...
+            array_keys($subprojects)
+        )) {
+            $BS          = 'BehaviouralSubproject';
+            $cleanString = htmlspecialchars($_REQUEST[$BS]);
+            $this->tpl_data['PreviousSubproject']
+                = Utility::getSubprojectsForProject()[$cleanString];
```
Comment (ridz1208) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: this call should pass a project, otherwise use getSubprojectList() for clarity

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 8]
```diff
@@ -545,19 +548,48 @@ class NDB_Form_Statistics extends NDB_Form
...
+                   'Name' => $subprojects[$currentSubProject],
+                  );
+            $subproject_query     = "AND (s.SubprojectID IS NULL";
+            $subproject_query    .= " OR s.SubprojectID=:spid)";
+            $this->params['spid'] =$currentSubProject;
         } else {
-            $suproject_query ='';
+            $subproject_query ="";
+        }
+        if (!in_array(
```
Comment (ridz1208) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: why is this in not on line 569 

i.e. why are you handling the `$_REQUEST['BehaviouralSubproject']` when it might not even be set !! it should be under the check on lines 565-568
```

if (isset($_REQUEST['BehaviouralSubproject'])
  && $_REQUEST['BehaviouralSubproject'] != ''
  && in_array($_REQUEST['BehaviouralSubproject'], array_keys($subprojects))
```

**Comment by sruthymathew123** [comment]
@ridz1208 I agree with Leo. The subproject filter is populating for the selected project in Behavioural/Reliability statistics. Also, I had a look on statistics count after applying the subproject filter whether there is any mismatch. No problems identified!

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 9]
```diff
@@ -545,19 +548,48 @@ class NDB_Form_Statistics extends NDB_Form
...
         }
-        $subprojList = substr($subprojList, 0, -1);
-        if (!empty($subprojList)) {
-            $suproject_query ="AND s.SubprojectID IN ($subprojList)";
+
+        $this->tpl_data['Subprojects'] = $subprojects;
+        $currentSubProject = null;
+        $subprojList       = substr($subprojList, 1, -1);
+        if (isset($_REQUEST['BehaviouralSubproject'])
+            && $_REQUEST['BehaviouralSubproject'] != ''
```
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Typically instead of `isset($var) && $var != ''` we use `!empty($var)`. This has the same truth value and has the benefit of being clearer and shorter.

Since this is one code convention we do actually have, please enforce it here 😅

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 10]
```diff
@@ -545,19 +548,48 @@ class NDB_Form_Statistics extends NDB_Form
...
+            && in_array($_REQUEST['BehaviouralSubproject'], array_keys($subprojects))
+        ) {
+            $currentSubProject = $_REQUEST['BehaviouralSubproject'];
```
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Instead:
```php
$currentSubProject = htmlspecialchars($_REQUEST['BehaviouralSubproject']);
```

Also I'd prefer `$currentSubproject` to `$currentSubProject` given that it's consistent with the other variable names.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 11]
```diff
@@ -545,19 +548,48 @@ class NDB_Form_Statistics extends NDB_Form
...
         } else {
-            $suproject_query ='';
+            $subproject_query ="";
+        }
+        if (!in_array(
+            $_REQUEST['BehaviouralSubproject'],
+            array_keys($subprojects)
+        )) {
+            $BS          = 'BehaviouralSubproject';
+            $cleanString = htmlspecialchars($_REQUEST[$BS]);
```
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: 🥇

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 12]
```diff
@@ -687,9 +719,9 @@ class NDB_Form_Statistics extends NDB_Form
         }
         $subprojList = substr($subprojList, 0, -1);
         if (!empty($subprojList)) {
```
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Instead, go for a construction like:

```php
$subproject_query = '';
if (!empty($subprojList)) {
    $subproject_query .= "AND s.SubprojectID IN ($subprojList)";
}
```

Basically no `else` statement. It's a bit cleaner and Dave will just **love** you for it.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 13]
```diff
@@ -1029,6 +1061,50 @@ class NDB_Form_Statistics extends NDB_Form
...
+
+        foreach ($subprojects as $key=>$val) {
+            $subprojList .= $key.",";
+        }
+
+        $this->tpl_data['Subprojects'] = $subprojects;
+        $currentSubProject = null;
+        $subprojList       = substr($subprojList, 1, -1);
+        if (isset($_REQUEST['ReliabilitySubproject'])
+            && $_REQUEST['ReliabilitySubproject'] != ''
```
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: `!empty` instead

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 14]
```diff
@@ -1029,6 +1061,50 @@ class NDB_Form_Statistics extends NDB_Form
...
+            && in_array($_REQUEST['ReliabilitySubproject'], array_keys($subprojects))
+        ) {
+            $currentSubProject = $_REQUEST['ReliabilitySubproject'];
```
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: `htmlspecialchars` around the `REQUEST`

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 9 above)
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: `!empty($var['x'])` doesn't have the same truth value as `isset($var['x']) && $var['x'] != '')`. It has the same truth value as `$var != ''`, but suppresses warnings if the index x isn't set. See https://www.virendrachandak.com/techtalk/php-isset-vs-empty-vs-is_null/ for a table of the truth values of the (somewhat ridiculous) number of ways PHP has to check if a variable's value is empty.

`$REQUEST['BehaviouralSubproject'] ?? ''`, on the other hand, does have the same truth value as `isset(x) && x`. It's how the `??` operator is defined in PHP (since PHP7..)

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 10 above)
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Subproject is consistently referred to with a lowercase p in most places in LORIS that I'm aware of.

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 7 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: also made all of these changes for the corresponding lines in the Reliability section

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 9 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: good to know!

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 10 above)
Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Yeah exactly

**Comment by ZainVirani** [comment]
@johnsaigle @ridz1208 @driusan changes made, will make more general (unrelated to this PR) security fixes in a new PR

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 15]
```diff
@@ -545,19 +545,56 @@ class NDB_Form_Statistics extends NDB_Form
         } else {
             $Param_Project = '';
         }
-        $subprojects = Utility::getSubprojectsForProject($currentProject);
+        $subprojects[null] = 'All Valid Sub-Projects';
+
+        foreach (Utility::getSubprojectsForProject($currentProject)
+            as $key=>$value) {
+            $subprojects[$key] = $value;
+        }
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: I think this could just be:

```
$subprojects = Utility::getSubprojectsForProject($currentProject)
$subprojects[null] = 'All Valid Subprojects'
```
but it's probably not worth taking the time to make the change.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 16]
```diff
@@ -545,19 +545,56 @@ class NDB_Form_Statistics extends NDB_Form
...
-            $suproject_query ='';
+            $subproject_query ="";
+        }
+        if ($_REQUEST['BehaviouralSubproject'] ?? ''
+            && !in_array(
+                $_REQUEST['BehaviouralSubproject'],
+                array_keys(
+                    $subprojects,
+                    true
+                )
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: I think I'm missing something.. isn't this just an if/else with the inverse logic of the if/else directly above it? Why not just add it to the existing code blocks?

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 17]
```diff
@@ -998,7 +1034,7 @@ class NDB_Form_Statistics extends NDB_Form
 
 
     /**
-     * ELIABILITY function
+     * RELIABILITY function
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: Are you sure this isn't supposed to be an e-liability function?

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 18]
```diff
@@ -1014,27 +1050,75 @@ class NDB_Form_Statistics extends NDB_Form
...
+        //SUBPROJECTS
+        $subprojects[null] = 'All Valid Sub-Projects';
+
+        foreach (Utility::getSubprojectsForProject($currentProject)
+            as $key=>$value) {
+            $subprojects[$key] = $value;
+        }
+
+        foreach ($subprojects as $key=>$val) {
+            $subprojList .= $key.",";
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: As far as I can tell, this is appending to a variable that doesn't exist at this point. I believe this will generate some PHP warnings.

I also don't see anywhere that this variable is used in the function, so this seems to be dead code that can just be removed.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 19]
```diff
@@ -1014,27 +1050,75 @@ class NDB_Form_Statistics extends NDB_Form
...
+        }
+
+        $this->tpl_data['Subprojects'] = $subprojects;
+        $currentSubproject = null;
+        $subprojList       = substr($subprojList, 1, -1);
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: same unused(?) variable.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 20]
```diff
@@ -1014,27 +1050,75 @@ class NDB_Form_Statistics extends NDB_Form
...
+        } else {
+            $subproject_query ="";
+        }
+        if ($_REQUEST['ReliabilitySubproject'] ?? ''
+            && !in_array(
+                $_REQUEST['ReliabilitySubproject'],
+                array_keys(
+                    $subprojects,
+                    true
+                )
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: same question about why this isn't in the existing if/else blocks, instead of adding a new (but opposite) if/else block.

File: modules/statistics/php/NDB_Form_statistics.class.inc [hunk 21]
```diff
@@ -1014,27 +1050,75 @@ class NDB_Form_Statistics extends NDB_Form
...
+            )) {
+            $RS          = 'ReliabilitySubproject';
+            $cleanString = htmlspecialchars($_REQUEST[$RS]);
+            $this->tpl_data['PreviousSubproject']
+                = Utility::getSubprojectList()[$cleanString];
+            $this->tpl_data['Warning']            = 1;
```
Comment (driusan) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: should this be `true` (a boolean) instead of `1` (an integer)?

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 16 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: good point, also there's an error in my `in_array(...)` call, `array_keys($subprojects, true)` doesn't make sense

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 17 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: RE: e-liability reliability

Comment (johnsaigle) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: 🤢

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 19 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: removed

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 16 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: disagree, there are some circumstances under which both else statements must be executed (its not perfectly inverse logic)

File: modules/statistics/php/NDB_Form_statistics.class.inc (hunk 21 above)
Comment (ZainVirani) [inline modules/statistics/php/NDB_Form_statistics.class.inc]: yes

**Comment by ridz1208** [comment]
@sruthymathew123 @suzanne-lee IBIS people this was intended for you guys (see redmine) please test

**Comment by driusan** [comment]
@ZainVirani Can you rebase this? (Hopefully that'll also cause it to get out of Travis not yet started purgatory.)

**Comment by ZainVirani** [comment]
@llevitis @sruthymathew123 rebased and tested it again, seems ok on my end but you may want to retest as well

**Comment by driusan** [comment]
@ZainVirani Conflicts, can you rebase?

**Comment by ZainVirani** [comment]
@driusan rebasing would be weird because the module was refactored, files have been split, moved, and filenames have changed. It may just be easier to redo the changes.

**Comment by driusan** [comment]
ok, if you're going to redo the changes can you close this PR then?

**Comment by ZainVirani** [comment]
Changes will be redone using refactored module
